Automatically link to source code of document.

samth opened the pull request:
Automatically link to source code of document.

This works, provided that:
  - the package is on the catalog at pkgs.racket-lang.org
  - the package is hosted on GitHub
  - the package source does not specify a branch

Restriction 3 is easy to lift with more parsing.
Restriction 2 could be lifted for other known sites (such as gitlab)
 or packages hosted as directories.
Restriction 1 would be harder to lift.

This only links to the _top level_ file that defines the overall
document, not the file defining the particular section. Fixing that
would require the Scribble renderer to provide more detailed
information, although it's certainly a possible extension.

Currently, some single-collection packages get the wrong answer,
but that will change when racket/pkg-index#25 finishes updating on
the server.

Closes #223
Closes #208
Closes racket/racket#874    
Related to #76, #228

mflatt commented:
This worked for me for the Plait documentation, but not for most things.

For most things, if I look in the consolev, I see an error about `source_found` being null. Probably the script should not fail if the URL doesn't match. Meanwhile, the things that failed to match have "?path=", such as `git://github.com/racket/scribble/?path=scribble-doc`, which seems not to match due to the `/` before the `?`.

(Is a regular expression really the right way to parse a URL in Javascript?)

samth commented:
Can you try again now? I've fixed a few bugs recently.

mflatt commented:
Yes, this works for me.

sorawee commented on scribble-lib/scribble/manual-racket.js:
```diff
@@ -15,14 +15,130 @@ AddOnLoad(function() {
+var cache = new Map();
+
+function ParseSource(source, mod_path, single_collection) {
+    let source_re = /^(https|git):\/\/\github.com\/([^\/]+)\/([^\?\/]+?)(\/|)(\.git|)(\?path=(.+)|)(#[A-Za-z0-9_-]*|)$/;
```
`let` doesn't work on old browsers. If you care about backward compatibility, you might want to use `var` instead.

sorawee commented on scribble-lib/scribble/manual-racket.js:
```diff
@@ -15,14 +15,130 @@ AddOnLoad(function() {
+
+    let parsed = source && mod_path && ParseSource(source, mod_path, single_collection);
+
+    let correct_url = `https://github.com/${parsed.user}/${parsed.repo}/tree/${parsed.branch}/${parsed.file_path}`;
```
Ditto for template literal.

samth commented:
@sorawee mentions old browser compatibility. I'm pretty sure both let and template literals are available for basically everyone who wants this feature, but maybe we should have a general policy about what we want to be compatible with.

mflatt commented:
That sounds fine for the new feature, as long as using potentially unsupported things doesn't cause a failure that breaks old features.

samth commented:
I think that both of these features would cause all of manual-racket.js to fail to parse.

samth commented on scribble-lib/scribble/manual-racket.js:
```diff
@@ -15,14 +15,130 @@ AddOnLoad(function() {
+        file_path = source_path + "/" + file_path;
+    }
+
+    return {user: source_user,
```
I think this should be fixed now. Can you check?

samth commented:
On the browser-compatibility side, here's [let](https://caniuse.com/#feat=let), [template literals](https://caniuse.com/#feat=template-literals), and [fetch](https://caniuse.com/#feat=fetch). All are over 95% of desktop browsers, and I expect even higher for people who want to edit Scribble files.

Having `let` or templates not work would cause the whole file to fail to parse. Having `fetch` not available would cause the code to error before doing anything useful (including for the previous message for how to link to the document in Scribble). Probably `fetch` could be handled with an exception handler or a callback. 

Additionally, the code uses `Map`, which also doesn't work in pre-2015 browsers.

None of these changes would make rendering the actual document break on any browser. If people want, I can take all of these out, though.

samth commented:
And about regexps -- using the `URL` constructor would take care of separating `git://github.com/foo/bar.git?path=xxx#master` into 

`URL { href: "git://github.com.com/foo/bar/?path=xxx#master", origin: "null", protocol: "git:", username: "", password: "", host: "", hostname: "", port: "", pathname: "//github.com.com/foo/bar/", search: "?path=xxx" }`

but that's not that helpful.

sorawee commented:
Just FYI: I would be very happy if Racket/Scribble switches to modern JS. I just raised the point in case you really want to commit to support very old browsers.

samth commented:
I just rewrote using the browser URL parser, and handled old browsers (I think; it's hard to test) and added GitLab support. @bennn can you test again?

sorawee commented on scribble-lib/scribble/manual-racket.js:
```diff
@@ -15,14 +15,175 @@ AddOnLoad(function() {
+        }
+        else {
+            fetch("https://pkgs.racket-lang.org/pkg/" + source_pkg + ".json")
+                .then(response => response.json())
```
It looks like you want to commit to support old browsers after all. In that case, arrow function here will cause a parsing error in old browser.

bennn commented:
Looks good for basic GitLab urls (pict-abbrevs) & ones with a path (csexpr)

I couldn't find any pkgs with usable version exceptions on GitLab

spdegabrielle commented:
This is awesome - Thank you.
